### DISCUSSION TITLE
density initialization at the left boundary

### DISCUSSION THREAD

**Discussion by Tissot11** [SmileiPIC/Smilei#778] [General]
I know that to use built-in function, one should have at lest 3 ghost cells (2nd oder shape function), but I though defining a density profile using Python function  should not have this limitation? I initialized a density profile using my own Python function, I was expecting the density to be 2 at `x=0` but it is 1. It gets 2 at the next point of `Probe` grid. 

![Figure 2024-12-07 185115 (27)](https://github.com/user-attachments/assets/3d7c422e-9f91-449d-a2de-c9975730366e)

I also define external field etc, using my Python profile and they appear to be correctly initialised at x=0, only the density profile is a bit weird...

**Comment by mccoys** [comment]
If your boundary conditions are not periodic, then the field interpolation will be done with no particles across the border. It is this acting as if there are no particles on the other side

**Comment by Tissot11** [comment]
Ok. Can one do something to make this better with `remove` boundary conditions? This blip in density is causing an unphysical filamentation that becomes important if I run the simulation on longer durations. A `x=0`, I am also injecting particles starting from `t=0`

**Comment by Tissot11** [comment]
Just a clarification, do you mean that interpolation in `Probe` diagnostic an issue or `python function parsing` in `Smilei`? Interpolation in `Probe` diagnostic is sometimes not good as I raised in #758 too. Do you mean that these issues at the boundary will always be present? if yes, could you please recommend something to alleviate this problem?

**Reply by mccoys** [reply]
It is not an issue with the code. The Probes interpolate density in a correct way for the pic algorithm. Your issue is I guess that there is no boundary  that behaves the way you want it. I am not sure exactly what would be necessary for you.

**Comment by Tissot11** [comment]
I am not sure if I agree with you on this. I repeated the same simulation with another PIC code (OSIRIS). Please see the figure, density is correctly initialised at the left boundary with `open` boundary conditions for `fields` and `particles`. There maybe a small blip at the boundary but not as strong as in Smilei where the value is the half of the value initialised.

![Figure 2024-12-09 131408](https://github.com/user-attachments/assets/949b9c5f-b718-4335-bcc4-48ab8e8a4e7d)

Small fluctuations can be expected, but values differing strongly like a factor of 2 in this case and factor of 3 in #758 are problematic...

**Reply by mccoys** [reply]
Did you check how density is measured in osiris? 

In smilei you can measure density with ParticleBinning, and it will show no decrease

**Comment by Tissot11** [comment]
So the results between `Probes` and `ParticleBinning` in Smilei would differ even if I choose same number of points for grids?

Currently, OSIRIS has something similar to `Probes` or `ParticleBinning` in Smilei and I am using similar number of points as I do for `Probes` in Smilei. 

What I would like to understand from you if this density blip at the boundary is just an artefact of the diagnostics being used? Or it is real and would be present  in PIC simulation cycle manifesting into some kind of plasma dynamics (filamentation in this case)?

**Comment by beck-llr** [comment]
Hello.
 The density drop on the boundary measured with Probes would be significantly mitigated with an interpolation order of one (this is maybe what happened when you ran OSIRIS ?). With an higher interpolation order (as is default in Smilei), almost half the stencil reaches cells which are outside the physical domain where the density is considered 0 hence the returned density is halved.
If you use a particle binning diagnostic with bins strictly located inside the physical domain you will not be affected by the interpolation scheme. So yes, the results will differ because you are not measuring exactly the same things.

So as you can see, the density can be different depending on how you measure it. Anyway this will never impact your simulation because the density is not used to solve Maxwell's equation. Charge density doesn't have any effect on the physics in an electromagnetic PIC simulation. It is here only for diagnostics purposes.

**Comment by Tissot11** [comment]
Thanks for the clear explanations, this helps me to understand the results! For me, physics is more important, so it is good to hear that it will not have impact on simulation.

I did use `quadratic` shape function in OSIRIS for a fair comparison with SMILEI, and not the default one `linear`.

**Comment by beck-llr** [comment]
In that case they must deposit density outside of the physical domain or use some interpolation trick on the boundary. Can't be sure.
In Smilei charge and current densities are zero outside of the physical domain and the Probe diagnostics are useful to check exactly what a particle would "see" with the chosen interpolation scheme. So, as I said, this drop in charge density has no influence but it might for some other quantities and this diagnostic is useful exactly for that.
